Draft blog post for Node.js WinRT projections (#598)
microsoft/winappCli: ifdef-blog-js-projections-draft.md

## How to add the projection to your Electron app

### Project setup

In your Electron app, install the CLI and initialize:

```bash
npm install --save-dev @microsoft/winappcli
npx winapp init . --use-defaults --add-js-bindings
```

`winapp` sets up the manifest and SDKs, adds `@microsoft/dynwinrt` + `@microsoft/dynwinrt-codegen` to your `package.json`, and writes typed bindings to `.winapp/bindings/`. For APIs that require package identity, like notifications and Phi Silica, run `npx winapp node add-electron-debug-identity` before starting Electron.

### Show a native Windows notification

The notification sample is the smallest. From the Electron main process:

```js
const {
  AppNotificationBuilder,
  AppNotificationManager,
} = require('./.winapp/bindings/index.js');

AppNotificationManager.default_.show(
  AppNotificationBuilder
    .create()
    .addText('Hello from Electron!')
    .addText('Powered by the Windows App SDK.')
    .buildNotification()
);

> ```

Reviewer: **This is the one place to show the gap vs. Electron's built-in notifications.**

Two `.addText()` lines is essentially what Electron's HTML5 `new Notification(title, { body })` already gives you on Windows — an Electron dev reading this may shrug. The important gap is what the built-in API *can't* do: **action buttons**, **inline text-box replies**, **progress bars**, **hero/inline images**, and reliable **Action Center persistence** under real package identity.

Suggest making the default example a richer toast so the reader immediately thinks "oh — that's the thing I could never do from Electron." e.g. add a button (and maybe an inline reply box):

```js
AppNotificationManager.default_.show(
  AppNotificationBuilder.create()
    .addText('Build finished')
    .addText('3 warnings, 0 errors')
    .addButton(AppNotificationButton.create('View log').addArgument('action', 'viewLog'))
    .buildNotification()
);
```

(Please confirm the exact generated binding names — `addButton` / `AppNotificationButton` / `addProgressBar` / `addTextBox` — against `.winapp/bindings`.) The plain two-liner can stay as the "hello world," but lead the reader to the buttons/inputs so the payoff is obvious.

Author: use addProgressBar to show more

Full walk-through in the [Show a notification from JavaScript](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-notification.md) guide.

### Run Phi Silica on-device AI

On-device AI is the same shape — no extra ceremony, no addon to maintain:

```js
const {
  AIFeatureReadyState, LanguageModel, TextSummarizer,
} = require('./.winapp/bindings/index.js');

if (LanguageModel.getReadyState() === AIFeatureReadyState.NotReady) {
  await LanguageModel.ensureReadyAsync();
}

const model = await LanguageModel.createAsync();
try {
  const op = TextSummarizer
    .createInstance(model)
    .summarizeParagraphAsync('Some long paragraph...');

  // Stream partial output as the model generates it
  op.progress((partial) => {
    process.stdout.write(partial);
  });

  const result = await op;
  console.log('\nDone:', result.text);
} finally {
  model.close();
}
```

Phi Silica needs a Copilot+ PC. Before running, add the `systemAIModels` restricted capability to `Package.appxmanifest`:

```xml
<Capabilities>
  <rescap:Capability Name="systemAIModels" />
</Capabilities>

> ```

Reviewer: **On adding `systemAIModels` — two parts:**

**1. Can `winapp` add it with a single command?** Not today. `winapp manifest` only exposes `generate` / `update-assets` / `add-alias`, so `systemAIModels` currently has to be hand-edited into the manifest (as shown). A `winapp manifest add-capability systemAIModels` (or a `winapp init --ai` opt-in) would collapse this to one step and make the guide cleaner — worth filing as a follow-up.

**2. Should `init` add it by default?** I'd lean **no**. `systemAIModels` is a *restricted* capability; declaring it on every `init` (including the majority of apps that never touch on-device AI) over-declares privilege and adds Store-certification justification for apps that don't need it. Note the default template already carries the `rescap` namespace + `runFullTrust`, so the namespace plumbing is free — but the capability itself should stay **opt-in via a single command**, not on-by-default. So the honest answer to "no harm to include by default?" is: there is mild harm (least-privilege + Store review), so prefer the one-command opt-in over defaulting it in.

Then refresh debug identity with `npx winapp node add-electron-debug-identity`. Full walk-through in the [Call Phi Silica from JavaScript](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-phi-silica.md) guide.

Running the snippet above in an Electron main process, the summary streams into the DevTools console chunk by chunk, followed by the final `Done:` line:

![Electron DevTools console: the Phi Silica summary streams in one partial chunk at a time via op.progress(), followed by a final "Done:" line with the complete summary.](./assets/phi-silica-console.gif)

The snippet above is the minimal shape. To see the projection driving a full app end-to-end, take a look at [**Electron on Windows Gallery**](https://github.com/microsoft/electron-on-windows-gallery) — an open-source sample gallery powered by this same setup, with samples for text generation, summarization, rewriting, OCR, image description, image scaling, and object extraction. Here are a few of the samples in action:

> ![Electron on Windows Gallery running several on-device AI samples end-to-end: text summarization, OCR, object remover, and image description — each driven by a handful of JavaScript against the generated bindings.](./assets/electron-gallery-samples.gif)

Reviewer: **Move this Gallery callout up — right after `### Project setup`.**

Framing the [Electron on Windows Gallery](https://github.com/microsoft/electron-on-windows-gallery) early sets the right expectation: this post walks through **2** scenarios inline (notifications + Phi Silica), but there are **many more** — OCR, image description, rewrite, object extraction, image scaling — that a reader can copy almost verbatim from the gallery. Leading with that tells them "the gallery is the full menu; these two inline samples are just the appetizer," instead of discovering it only at the very end. We can keep a short line here saying "we'll show 2 of them in detail below."

## Extending to Windows SDK and beyond

By default `winapp` feeds codegen the supported Windows App SDK WinRT surface (UI-only packages like `Microsoft.WindowsAppSDK.WinUI` and `Microsoft.Web.WebView2` are excluded — see the scope note below). To pull in Windows SDK classes (for example to open a `FileOpenPicker`, decode an image with `BitmapDecoder`, or talk to `Windows.Web.Http.HttpClient`), add them to `package.json`:

```jsonc
{
  "winapp": {
    "jsBindings": {
      "additionalWinmds": [
        { "namespace": "Windows.Storage",          "classes": ["StorageFile"] },

>         { "namespace": "Windows.Graphics.Imaging", "classes": ["BitmapDecoder"] }

Reviewer: **`StorageFile` is a weak hook — swap it for the native Share experience.**

`StorageFile` doesn't answer "why do I care." A far more resonant showcase is the **native Windows Share sheet** via `Windows.ApplicationModel.DataTransfer.DataTransferManager` (`DataRequested` + `ShowShareUI`) — one call pops the system share flyout (Mail, Teams, nearby share, …), which Electron devs normally hand-roll a custom UI to fake. That's a genuine "wow, I get the OS share sheet for free" moment.

Suggest making this the flagship "Extending to Windows SDK" example and **adding a screenshot** of the Windows Share flyout (like the toast/console screenshots above).

One caveat to spell out: for desktop/Win32 apps `ShowShareUI` needs the app window handle (`IDataTransferManagerInterop.ShowShareUIForWindow`), so a sentence on wiring the HWND from JS would help. If that interop is awkward in the current bindings, `Clipboard` or `FileOpenPicker` are simpler fallback crowd-pleasers — but Share is the strongest story.

Author: Agree Share is a stronger story and a great follow-up showcase. I tried the generated  DataTransferManager  path locally; for Electron/Win32 the plain  `DataTransferManager.showShareUI` API goes through the UWP/current-view path and fails. The right path is the HWND interop API, `IDataTransferManagerInterop.ShowShareUIForWindow(hwnd)`. That interface is not part of the normal WinRT .winmd  surface we generate from — it’s a Win32/COM interop interface from the SDK headers — so it doesn’t show up in the generated bindings today. 

I will use Clipboard scenario to replace it

      ]
    }
  }
}
```

Then `npx winapp node generate-bindings`. The codegen transitively pulls in dependent types — you only list the entry-point classes.

For third-party WinRT components, or to include a package `winapp` doesn't ship by default, point the entry directly at a `.winmd` file: `{ "winmdPath": "path/to/Foo.winmd", "namespace": "Foo.Bar", "classes": ["Baz"] }`.

## Plain Node.js: a dev-mode quick-start

> Not on Electron? The same projections work from a plain Node.js process too. For local prototyping — trying a WinRT API on your own box without shipping an MSIX — register a loose-layout package and give it a command-line alias:

Reviewer: **This section tries to cover several situations at once and runs long** (loose-layout register + copy `node.exe` + execution alias + `run --no-launch` + `unregister`). We agreed we can't show them all inline.

Pick the **single simplest + most powerful** path as the one hero example — I'd keep the execution-alias + `winapp run` flow — and move the other variations (bare `winapp run` without an alias, the copy-node-into-project trick, packaging-later notes) into a dedicated guide, e.g. `docs/guides/node/dev-mode.md`, and link to it from here. Keeps the blog tight and gives the extra detail a proper home.

# Introducing a dynamic WinRT projection for Node.js

> 

Reviewer 2: When you go to publish this in word press, make sure to add a nice header image as well. Something similar to this: https://devblogs.microsoft.com/ifdef-windows/wp-content/uploads/sites/61/2026/05/D__win-dev-skills_docs_assets_blog-banner.html-2.webp

To make these we usually tell copilot to make a graphic using html and then we screenshot the html to get a png you can add to the post.

## What's in the box

Three npm packages, designed to be used together:

- **[`@microsoft/dynwinrt`](https://www.npmjs.com/package/@microsoft/dynwinrt)** — the runtime that powers the generated bindings, dispatching WinRT calls at execution time. Prebuilt for x64 and arm64 Windows, so it installs from npm like any other package — no compiler, no `node-gyp`.
- **[`@microsoft/dynwinrt-codegen`](https://www.npmjs.com/package/@microsoft/dynwinrt-codegen)** — the code generator that reads `.winmd` metadata and emits typed `.js` + `.d.ts`.
- **[`@microsoft/winappcli`](https://www.npmjs.com/package/@microsoft/winappcli)** — the Windows App Development CLI. It manages the NuGet packages that ship WinRT metadata, runs codegen, pins the matching `@microsoft/dynwinrt` runtime, and handles debug package identity.

You only install `@microsoft/winappcli`; it brings in the other two for you.

## What you can build

In this post we'll walk through two common Windows API scenarios from JavaScript:

- **Native notifications** with `AppNotificationBuilder` / `AppNotificationManager`
- **Phi Silica on-device AI** with `LanguageModel` / `TextSummarizer` (Copilot+ PCs)

> Both are written entirely in JavaScript. The same pattern extends to file pickers, WinML execution-provider discovery, and other WinRT surfaces — full walk-throughs live in the [Electron guides](https://github.com/microsoft/winappCli/tree/main/docs/guides/electron).

Reviewer 2: I would remove the second sentence in this paragraph and instead at the bottom on this walkthrough section add a sentence that says "For more examples of how to use the dynamic WinRT projections, see [Electron Guides].

> For the full menu of what this unlocks, take a look at [**Electron on Windows Gallery**](https://github.com/microsoft/electron-on-windows-gallery) — an open-source sample gallery powered by this same setup, with samples for text generation, summarization, rewriting, OCR, image description, image scaling, and object extraction.

Reviewer 2: I would also move this paragraph to later in the document. Keep this section focused on walking through the 2 examples you listed at the top

Author: I moved it into a separate paragraph so it’s no longer part of the two examples, but I kept it before the samples because I feel the demos are one of the most engaging parts of the post and can help capture readers’ interest before they dive into the examples.

```js
const {
  AppNotificationBuilder,
  AppNotificationManager,
  AppNotificationProgressBar,
} = require('./.winapp/bindings/index.js');

const progress = AppNotificationProgressBar
  .create()
  .setTitle('Processing with Windows AI')
  .setStatus('Running locally')
  .setValue(0.65)
  .setValueStringOverride('65%');

AppNotificationManager.default_.show(
  AppNotificationBuilder
    .create()
    .addText('Windows AI task running')
    .addText('This notification uses a native Windows progress bar.')
    .addProgressBar(progress)
    .buildNotification()
);
```
Run it, and Windows fires the same native toast you'd get from a C#/C++ app — including the progress bar:

![Windows toast notification from an Electron app titled "test-electron-app", reading "Windows AI task running" and showing a 65% progress bar labeled "Processing with Windows AI."](./assets/electron-toast-hello-from-electron.png)

Full walk-through in the [Show a notification from JavaScript](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-notification.md) guide.

### Run Phi Silica on-device AI

> On-device AI is the same shape — no extra ceremony, no addon to maintain:

Reviewer 2: Would change "same shape — no extra ceremony" to "You can also add on-device AI to Electron directly from your app's JavaScript: "

```js
const {
  AIFeatureReadyState, LanguageModel, TextSummarizer,
} = require('./.winapp/bindings/index.js');

if (LanguageModel.getReadyState() === AIFeatureReadyState.NotReady) {
  await LanguageModel.ensureReadyAsync();
}

const model = await LanguageModel.createAsync();
try {
  const op = TextSummarizer
    .createInstance(model)
    .summarizeParagraphAsync('Some long paragraph...');

  // Stream partial output as the model generates it
  op.progress((partial) => {
    process.stdout.write(partial);
  });

  const result = await op;
  console.log('\nDone:', result.text);
} finally {
  model.close();
}
```

> Phi Silica needs a Copilot+ PC. Before running, add the `systemAIModels` restricted capability to `Package.appxmanifest`:

Reviewer 2: When you move this draft to wordpress I would add some formatting to this line so it's super visible to the users that they need a Copilot+PC for this example to work .

When you're done experimenting, unregister the development package:

```powershell
npx winapp unregister
```

> **This is a dev-mode convenience.** `winapp run` registers a loose-layout package on your local machine so you can iterate without building an MSIX. When you're ready to distribute, package and sign the same app layout with `winapp pack` + `winapp sign`; the generated bindings and JavaScript code carry forward.

## How it works, briefly

`dynwinrt-codegen` runs during `winapp init` / `restore` / `generate-bindings` and turns `.winmd` metadata into typed JavaScript wrappers — no native code is generated per class. At execution time, `@microsoft/dynwinrt` invokes the underlying COM vtables directly, handling the WinRT plumbing (HSTRINGs, HRESULTs → JavaScript exceptions, async operations → Promises, collections, structs, enums, delegates) transparently.

The projection targets non-UI WinRT APIs such as AI, storage, notifications, networking, and similar system capabilities — not UI hosting like XAML / WinUI or WebView2.

For the full design, see [`dynwinrt/design.md`](https://github.com/microsoft/dynwinrt/blob/main/design.md).

## Final thoughts

> This is public preview, and there's a lot we still want to sharpen. If a WinRT class doesn't shape well in JS, a TypeScript type feels off, or a scenario you care about isn't covered yet, please file feedback — we're actively deciding what to invest in next.

Reviewer 2: Add a link to the repo for where you want folks to file issues.

## Resources and feedback

> ### winapp CLI

Reviewer 2: Make sure to have all "WinApp CLI" instances in the doc capitalize the "W" and "A" unless you are describing a call to the "winapp" command

- Repo and full command reference: [`microsoft/winappCli`](https://github.com/microsoft/winappCli)
- Install: [`@microsoft/winappcli`](https://www.npmjs.com/package/@microsoft/winappcli) on npm
- Electron getting-started guides: [setup](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/setup.md) · [file picker](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-file-picker.md) · [notification](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-notification.md) · [Phi Silica](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-phi-silica.md) · [WinML](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-winml.md)
- Debug identity for Electron: [Setup guide → Understanding Debug Identity](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/setup.md#step-5-understanding-debug-identity)

> - File bugs / feature requests: [winappCli issues](https://github.com/microsoft/winappCli/issues)

Reviewer 2: Might be good to distinguish somewhere when developers should file feedback under the winappcli repo vs dynwinrt repo. So devs know which repo is responsible for what parts of the product

### dynwinrt (the projection itself)

> - Install: [`@microsoft/dynwinrt`](https://www.npmjs.com/package/@microsoft/dynwinrt) · [`@microsoft/dynwinrt-codegen`](https://www.npmjs.com/package/@microsoft/dynwinrt-codegen) on npm

Reviewer 2: Is there a reason why someone would install the projections package directly instead of through winapp?

> # Introducing a dynamic WinRT projection for Node.js

Reviewer 3: Main feedback is to make the title understandable by audience that is not in the Microsoft/Windows world. WinRT doesn't mean much to that audience. Lead with the value from their viewpoint. Something like "New way to call native Windows APIs from JavaScript in Node.js- introducing dynamic API projections". Same for the banner image

![Dynamic WinRT projection for Node.js — public preview banner showing the tagline "Call Windows Runtime APIs directly from JavaScript in Electron and Node.js" alongside a code snippet using LanguageModel and TextSummarizer from the generated bindings.](./assets/blog-banner.png)

> We've been working on something new for Electron developers on Windows: **a dynamic WinRT projection for Node.js** that lets your Electron app — or any Node.js process — call many non-UI WinRT APIs (APIs described by `.winmd` metadata in the Windows App SDK or Windows SDK) directly from JavaScript or TypeScript. No app-specific native addon, no `node-gyp` build step, no C++ or C# wrapper in your project. 🚀

Reviewer 4: I am viewing the first couple of paragraphs of this blog post and my eyes were immediately drawn to a few sets of em dashes (—) which we know to be a common AI artifact. I would hate for our readers to discount this content because of this. Is it possible to rewrite these parts to have a more human voice?

> We've been working on something new for Electron developers on Windows: **a dynamic WinRT projection for Node.js** that lets your Electron app — or any Node.js process — call many non-UI WinRT APIs (APIs described by `.winmd` metadata in the Windows App SDK or Windows SDK) directly from JavaScript or TypeScript. No app-specific native addon, no `node-gyp` build step, no C++ or C# wrapper in your project. 🚀

Reviewer 3: ```suggestion
We've been working on something new for Electron developers on Windows: **a dynamic Windows Runtime API (WinRT) projections for Node.js** that lets your Electron app - or any Node.js process - call non-UI WinRT APIs (APIs described by `.winmd` metadata in the Windows App SDK or Windows SDK) directly from JavaScript or TypeScript. No app-specific native addon, no `node-gyp` build step, no C++ or C# wrapper in your project. 🚀
```

> We've been working on something new for Electron developers on Windows: **a dynamic WinRT projection for Node.js** that lets your Electron app — or any Node.js process — call many non-UI WinRT APIs (APIs described by `.winmd` metadata in the Windows App SDK or Windows SDK) directly from JavaScript or TypeScript. No app-specific native addon, no `node-gyp` build step, no C++ or C# wrapper in your project. 🚀

Reviewer 3: Might be worth adding a sentence or two to introduce the motivation of why developers should care in this intro, what is the pain point we are solving?

## How this projection is different

> 

Reviewer 4: It's mentioned in the subheader and in the sentence before the list that this projection takes a different approach, and I am wondering why it is taking a different approach?

> C++/WinRT, C#/WinRT, `windows-rs`, and PyWinRT project the Windows Runtime into their respective languages. This adds JavaScript to that list — targeting the **Node.js runtime** — but takes a different approach:

Reviewer 4: ```suggestion
C++/WinRT, C#/WinRT, `windows-rs`, and PyWinRT project the Windows Runtime into their respective languages. This projection adds JavaScript to that list — targeting the **Node.js runtime** — but takes a different approach:
```

> C++/WinRT, C#/WinRT, `windows-rs`, and PyWinRT project the Windows Runtime into their respective languages. This adds JavaScript to that list — targeting the **Node.js runtime** — but takes a different approach:

Reviewer 3: Add links for these for the user

- **Regenerate, don't rebuild.** New `.winmd` metadata ships → rerun codegen → updated bindings. No per-SDK native wrapper, no waiting on hand-authored bindings.

## See it in action

We've already used this projection to power [**Electron on Windows Gallery**](https://github.com/microsoft/electron-on-windows-gallery) — an open-source Electron app that showcases the range of native Windows functionality reachable from Electron. It ships interactive samples covering the current Windows on-device AI APIs — text generation, summarization, rewriting, text-to-table, image description, OCR, image scaling, object extraction, and object removal — alongside JavaScript sample code, API documentation, and getting-started guides for building your own Electron-on-Windows features.

<p align="center">
  <img src="./assets/electron-gallery-samples.gif" width="720" alt="Electron on Windows Gallery running several on-device AI samples end-to-end: text summarization, OCR, object remover, and image description — each driven by a handful of JavaScript against the generated bindings." />
</p>

> ## What you can build

Reviewer 4: This header makes it sound like you're going to tell me the full capabilities of this new projection but it's actually just setting up the two demo's that are going to be shared in this post.
```suggestion
## Examples of what you can build
```
Also, maybe this header should be after "What's in the box" so it flows better into the demo walkthroughs.

In this post we'll walk through two common Windows API scenarios from JavaScript:

> - **Native notifications** with `AppNotificationBuilder` / `AppNotificationManager`

Reviewer 3: I love these examples, add more so it's not just the obvious ones, many APIs in the sdks

- **Phi Silica on-device AI** with `LanguageModel` / `TextSummarizer` (Copilot+ PCs)

Both are written entirely in JavaScript.

## What's in the box

Three npm packages, designed to be used together:

- **[`@microsoft/dynwinrt`](https://www.npmjs.com/package/@microsoft/dynwinrt)** — the shared native runtime that dispatches WinRT calls at execution time. Prebuilt for x64 and arm64 Windows.
- **[`@microsoft/dynwinrt-codegen`](https://www.npmjs.com/package/@microsoft/dynwinrt-codegen)** — the code generator that reads `.winmd` metadata and emits JavaScript bindings with TypeScript types (`.js` + `.d.ts`).
- **[`@microsoft/winappcli`](https://www.npmjs.com/package/@microsoft/winappcli)** — the Windows App Development CLI. It manages the NuGet packages that ship WinRT metadata, runs codegen, pins the matching `@microsoft/dynwinrt` runtime, and handles debug package identity.

You only install `@microsoft/winappcli`; it brings in the other two for you.

## How to add the projection to your Electron app

### Project setup

In your Electron app, install the CLI and initialize:

```bash
npm install --save-dev @microsoft/winappcli
npx winapp init . --use-defaults --add-js-bindings
```

> The WinApp CLI sets up the manifest and SDKs, adds `@microsoft/dynwinrt` + `@microsoft/dynwinrt-codegen` to your `package.json`, and writes typed bindings to `.winapp/bindings/`. Then run this once before starting Electron:

Reviewer 4: ```suggestion
The WinApp CLI sets up the manifest and SDKs, adds `@microsoft/dynwinrt` + `@microsoft/dynwinrt-codegen` to your `package.json`, and writes typed bindings to `.winapp/bindings/`. Run this once before starting Electron:
```

> The WinApp CLI sets up the manifest and SDKs, adds `@microsoft/dynwinrt` + `@microsoft/dynwinrt-codegen` to your `package.json`, and writes typed bindings to `.winapp/bindings/`. Then run this once before starting Electron:

Reviewer 3: It's also worth mentioning that the bindings and output can be configured in package.json and that the config object is written on the init command

```bash

> npx winapp node add-electron-debug-identity

Reviewer 3: This is only needed for APIs that need package identity, not all do - would be good to clarify and link to the docs/guides for how this works.

```

### Show a rich Windows notification

Electron can already show a basic notification, but Windows App SDK notifications let you use richer Windows-native toast features like progress bars, actions, inputs, and scenarios. Here's a progress notification from the Electron main process:

```js
const {
  AppNotificationBuilder,
  AppNotificationManager,
  AppNotificationProgressBar,
} = require('./.winapp/bindings/index.js');

const progress = AppNotificationProgressBar
  .create()
  .setTitle('Processing with Windows AI')
  .setStatus('Running locally')
  .setValue(0.65)
  .setValueStringOverride('65%');

AppNotificationManager.default_.show(
  AppNotificationBuilder
    .create()
    .addText('Windows AI task running')
    .addText('This notification uses a native Windows progress bar.')
    .addProgressBar(progress)
    .buildNotification()
);
```

> Run it, and Windows fires the same native toast you'd get from a C#/C++ app — including the progress bar:

Reviewer 4: ```suggestion
Run the app, and Windows fires the same native toast you'd get from a C#/C++ app — including the progress bar:
```

```js
const {
  AIFeatureReadyState, LanguageModel, TextSummarizer,
} = require('./.winapp/bindings/index.js');

if (LanguageModel.getReadyState() === AIFeatureReadyState.NotReady) {
  await LanguageModel.ensureReadyAsync();
}

const model = await LanguageModel.createAsync();
try {
  const op = TextSummarizer
    .createInstance(model)
    .summarizeParagraphAsync('Some long paragraph...');

  // Stream partial output as the model generates it
  op.progress((partial) => {
    process.stdout.write(partial);
  });

  const result = await op;
  console.log('\nDone:', result.text);
} finally {
  model.close();
}
```

Before running, add the `systemAIModels` restricted capability to `Package.appxmanifest`:

```xml
<Capabilities>
  <rescap:Capability Name="systemAIModels" />
</Capabilities>
```

Then refresh debug identity:

```bash
npx winapp node add-electron-debug-identity
```

Full walk-through in the [Call Phi Silica from JavaScript](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-phi-silica.md) guide.

Running the snippet above in an Electron main process, the summary streams into the DevTools console chunk by chunk, followed by the final `Done:` line:

> ![Electron DevTools console: the Phi Silica summary streams in one partial chunk at a time via op.progress(), followed by a final "Done:" line with the complete summary.](./assets/phi-silica-console.gif)

Reviewer 4: Moved the reference to the guide to the end to 1. make sure people see the working example and 2. make things consistent with the other walkthrough above.
```suggestion
Running the snippet above in an Electron main process, the summary streams into the DevTools console chunk by chunk, followed by the final `Done:` line:

![Electron DevTools console: the Phi Silica summary streams in one partial chunk at a time via op.progress(), followed by a final "Done:" line with the complete summary.](./assets/phi-silica-console.gif)

Full walk-through in the [Call Phi Silica from JavaScript](https://github.com/microsoft/winappCli/blob/main/docs/guides/electron/js-phi-silica.md) guide.
```
